### PR TITLE
[WIP] Add array type interface

### PR DESCRIPTION
**Pull request by matthiasbeyer**

### CODE REVIEW

File: src/objects/array.h [hunk 1]
```diff
@@ -28,7 +28,217 @@
...
+ * ws_array type definition
+ *
+ * The ws_array type implementation operates on `void*` internally.
+ */
+struct ws_array {
+    struct ws_object obj; //!< Supertype
+
+    struct {
+        size_t len;
+        pthread_rwlock_t rwl;
```
Comment (neithernut) [inline src/objects/array.h]: Please use _one_ rw_lock for the entire array.

File: src/objects/array.h [hunk 2]
```diff
@@ -28,7 +28,206 @@
...
+    struct ws_array* const self //!< Array object
+);
+
+/**
+ * Get number of used elements of an array object
+ *
+ * @return Number of used elements in the array, zero on NULL passed
+ */
+size_t
+ws_array_nused(
```
Comment (neithernut) [inline src/objects/array.h]: If the array would resize automatically (double the size of the allocated space if you run out of memory), we could ommit this (and just use `len`).
Manual `resizes could also be omitted`.

File: src/objects/array.h [hunk 3]
```diff
@@ -28,7 +28,206 @@
...
+ *
+ * @warning stops iteration whenever `iter` returns false
+ *
+ * @return true if the iteration exited normally (over elements was iterated),
+ * else false
+ */
+bool
+ws_array_foreach(
+    struct ws_array* const self, //!< Array object
+    bool (*iter)(void* entry, void* etc), //!< Iteration callback
```
Comment (neithernut) [inline src/objects/array.h]: context comes first.

**Comment by matthiasbeyer** [comment]
Discussion: I save `void*`s here. This prevents me from doing ref-counting automatically. The spec says, I should save `struct ws_object` instances, but I'd prefer this implementation.

We _could_ add a subclass for `struct ws_array`, `struct ws_object_array` for example, which does refcounting then. I guess this would be somewhat cleaner.

File: src/objects/array.h [hunk 4]
```diff
@@ -28,7 +28,167 @@
...
+ *
+ * @warning stops iteration whenever `iter` returns false
+ *
+ * @return true if the iteration exited normally (over elements was iterated),
+ * else false
+ */
+bool
+ws_array_foreach(
+    struct ws_array* const self, //!< Array object
+    bool (*iter)(void* etc, void* entry), //!< Iteration callback
```
Comment (neithernut) [inline src/objects/array.h:159]: _Maybe_ rename `etc` -> `ctx`.

**Comment by neithernut** [comment]
Ok, we may want some subtypes (e.g. one for object, one or more for value types).
But for now, this is sufficient.

File: src/objects/array.h (hunk 4 above)
Comment (matthiasbeyer) [inline src/objects/array.h:159]: You can also pass non-context stuff, so I'd like to keep this.

**Comment by matthiasbeyer** [comment]
@neithernut ACK for squashing?

**Comment by neithernut** [comment]
ACK.

**Comment by matthiasbeyer** [comment]
Rebased. Feel free to review again.

**Comment by neithernut** [comment]
ACK.

**Pull request merged**